### PR TITLE
Ignore some metadata modificaiton events

### LINKED ISSUE #176: Lorri starts nix-builds in a loop
I've started using lorri from tip of master because of the SSL issues in earlier builds and some other fixes. Since then I've seen my cpu usage spike a lot. Afaict it is lorri, launching nix-builds over and over again. This happens very often, probably several times an hour. I end up restarting lorri which gets rid of those builds (for a while), here's an excerpt from the systemd journal:

```
-- Logs begin at Sat 2019-04-27 23:44:07 CEST. --
Oct 16 14:01:47 europa lorri[923]:     BuildResults {
Oct 16 14:01:47 europa lorri[923]:         output_paths: OutputPaths {
Oct 16 14:01:47 europa lorri[923]:             shell_gc_root: RootPath(
Oct 16 14:01:47 europa lorri[923]:                 "/home/john/.cache/lorri/gc_roots/3f193d3e7cddce96a26099d2979cb202/gc_root/shell_gc_root"
Oct 16 14:01:47 europa lorri[923]:             )
Oct 16 14:01:47 europa lorri[923]:         }
Oct 16 14:01:47 europa lorri[923]:     }
Oct 16 14:01:47 europa lorri[923]: )
Oct 16 14:01:47 europa lorri[923]: Started
Oct 16 14:01:47 europa nix-daemon[8151]: accepted connection from pid 15039, user john (trusted)
Oct 16 14:01:48 europa lorri[923]: Completed(
Oct 16 14:01:48 europa lorri[923]:     BuildResults {
Oct 16 14:01:48 europa lorri[923]:         output_paths: OutputPaths {
Oct 16 14:01:48 europa lorri[923]:             shell_gc_root: RootPath(
Oct 16 14:01:48 europa lorri[923]:                 "/home/john/.cache/lorri/gc_roots/3f193d3e7cddce96a26099d2979cb202/gc_root/shell_gc_root"
Oct 16 14:01:48 europa lorri[923]:             )
Oct 16 14:01:48 europa lorri[923]:         }
Oct 16 14:01:48 europa lorri[923]:     }
Oct 16 14:01:48 europa lorri[923]: )
Oct 16 14:01:48 europa lorri[923]: Started
Oct 16 14:01:48 europa nix-daemon[8151]: accepted connection from pid 15052, user john (trusted)
Oct 16 14:01:49 europa lorri[923]: Completed(
Oct 16 14:01:49 europa lorri[923]:     BuildResults {
Oct 16 14:01:49 europa lorri[923]:         output_paths: OutputPaths {
Oct 16 14:01:49 europa lorri[923]:             shell_gc_root: RootPath(
Oct 16 14:01:49 europa lorri[923]:                 "/home/john/.cache/lorri/gc_roots/3f193d3e7cddce96a26099d2979cb202/gc_root/shell_gc_root"
Oct 16 14:01:49 europa lorri[923]:             )
Oct 16 14:01:49 europa lorri[923]:         }
Oct 16 14:01:49 europa lorri[923]:     }
Oct 16 14:01:49 europa lorri[923]: )
Oct 16 14:01:49 europa lorri[923]: Started
Oct 16 14:01:49 europa nix-daemon[8151]: accepted connection from pid 15065, user john (trusted)
Oct 16 14:01:49 europa lorri[923]: Completed(
Oct 16 14:01:49 europa lorri[923]:     BuildResults {
Oct 16 14:01:49 europa lorri[923]:         output_paths: OutputPaths {
Oct 16 14:01:49 europa lorri[923]:             shell_gc_root: RootPath(
Oct 16 14:01:49 europa lorri[923]:                 "/home/john/.cache/lorri/gc_roots/3f193d3e7cddce96a26099d2979cb202/gc_root/shell_gc_root"
Oct 16 14:01:49 europa lorri[923]:             )
```

### PR DESCRIPTION
**Pull request by theotherjimmy**
We ignore metadata modificaiton events for the profiles directory
tree as it is a symlink forest that is used to keep track of
channels and nix will unconditionally update the metadata of each
link in this forest.
These ModifyKind::Metadata events are spurious and they can easily
cause a rebuild-loop when the shell.nix files  does not pin it's
version of nixpkgs or other channels.
When a Nix channel is updated we will recieve a data modification
event, so ignoring these metadata modifications will not impact
lorri's ability to correctly watch for changes.

Fixes #176

### CODE REVIEW

**Comment by curiousleo** [comment]
This seems to be a pragmatic solution to the issue!

Two comments:
- Please run `cargo fmt`, this should fix the CI failure. You can run CI locally using the `ci` command provided by the top-level `shell.nix` to check if everything is in order.
- Please add a comment explaining _why_ metadata events in that particular directory are ignored, perhaps even including a link to https://github.com/NixOS/nix/blob/629b9b0049363e091b76b7f60a8357d9f94733cc/src/libstore/local-store.cc#L74-L80 to be really precise.

**Comment by theotherjimmy** [comment]
Yeah, I forgot to run `cargo fmt`, my bad. I'll put most of the commit message into a comment above the code in question.

**Comment by theotherjimmy** [comment]
@curiousleo I have updated the commit to fix the formatting and include a comment explaining why we ignore the profiles dir and linking to the code, as I like being really precise.

**Comment by curiousleo** [comment]
> include a comment explaining why we ignore the profiles dir and linking to the code

That sounds good! I can't see that comment in the current state of the PR though. Perhaps that commit still needs to be pushed?

**Comment by curiousleo** [comment]
I'm about to sign off for the next couple of weeks. Don't take any lack of activity on this PR personally, please :)

**Comment by theotherjimmy** [comment]
Yep. Sorry about that. I see it on github now.

File: src/watch.rs [hunk 1]
```diff
@@ -152,8 +153,29 @@ impl Watch {
...
     }
 
-    fn path_is_interesting(&self, path: &PathBuf) -> bool {
+    fn path_is_interesting(&self, path: &PathBuf, kind: &EventKind) -> bool {
         path_match(&self.watches, path)
+            && match kind {
+                // We ignore metadata modification events for the profiles directory
+                // tree as it is a symlink forest that is used to keep track of
+                // channels and nix will uconditionally update the metadata of each
+                // linnk in this forest. See https://github.com/NixOS/nix/blob/629b9b0049363e091b76b7f60a8357d9f94733cc/src/libstore/local-store.cc#L74-L80
```
Comment (curiousleo) [inline src/watch.rs]:
```suggestion
                // link in this forest. See https://github.com/NixOS/nix/blob/629b9b0049363e091b76b7f60a8357d9f94733cc/src/libstore/local-store.cc#L74-L80
```

File: src/watch.rs [hunk 2]
```diff
@@ -152,8 +153,29 @@ impl Watch {
...
+                // for the unconditional update. These metadata mofication events are
```
Comment (curiousleo) [inline src/watch.rs]:
```suggestion
                // for the unconditional update. These metadata modification events are
```

File: src/watch.rs [hunk 3]
```diff
@@ -152,8 +153,29 @@ impl Watch {
...
+                // spurious annd they can easily cause a rebuild-loop when a shell.nix
+                // file does not pin it's version of nixpkgs or other channels. When
```
Comment (curiousleo) [inline src/watch.rs]:
```suggestion
                // file does not pin its version of nixpkgs or other channels. When
```

File: src/watch.rs [hunk 4]
```diff
@@ -152,8 +153,29 @@ impl Watch {
...
+                // a Nix channel is updated we recieve many other types of events, so
```
Comment (curiousleo) [inline src/watch.rs]:
```suggestion
                // a Nix channel is updated we receive many other types of events, so
```

**Comment by theotherjimmy** [comment]
Gotta get comment spellcheck setup. Thanks for the fixes.

**Comment by Profpatsch** [comment]
Looking absolutely fabulous! Thanks for this patch.

**Pull request merged**